rfd/0028-cluster-config-resources.md

```diff
@@ -0,0 +1,178 @@
+    int64 KeepAliveInterval;
+
+    // KeepAliveCountMax is the number of keep-alive messages that can be missed
+    // before
+    // the server disconnects the connection to the client.
+    int64 KeepAliveCountMax;
+
+    // LocalAuth is true if local authentication is enabled.
+    bool LocalAuth;
+
+    // SessionControlTimeout is the session control lease expiry and defines
+    // the upper limit of how long a node may be out of contact with the auth
+    // server before it begins terminating controlled sessions.
+    int64 SessionControlTimeout;
+}
+```
+
+It is proposed to distribute them in the following fashion:
+
+```proto
+message ClusterMeta {
+    // This resource is not genuine configuration, it would contain only the
+    // common resource metadata.  The metadata name field would be used to
+    // represent the ClusterID.
+}
+
+message SessionRecordingConfig {
+    string SessionRecording;
+    string ProxyChecksHostKeys;
+}
+
+message ClusterNetworkingConfig {
+    int64 ClientIdleTimeout;
+    int64 KeepAliveInterval;
+    int64 KeepAliveCountMax;
+    int64 SessionControlTimeout;
+}
+
+// The already existing AuditConfig is turned into a standalone resource.
+message AuditConfig { ... }
+
+// DisconnectExpiredCert & LocalAuth fields are moved into ClusterAuthPreference.
+message ClusterAuthPreference {
+    ...
+    bool AllowLocalAuth;
+    bool DisconnectExpiredCert;
+}
+```
+
+#### Data migration & docs updates
+
+A new step should be added to `migrateLegacyResources` (called during auth
+server initialization) that takes care of migrating the `ClusterConfig` data
+stored in the backend to the new resource structures.  This probably implies
+that `ClusterConfig` as a structure needs to be kept around for a couple of
+subsequent Teleport versions for marshalling purposes.
+
+Documentation should be updated accordingly.
+
+### Restricting to a subset of values of a field
+
+Some users should not be capable of setting a field to any of the otherwise
+accepted values.  It is proposed to satisfy this requirement by introducing new
+RBAC verbs, interpreted as implying additional constraints. Such new verbs
+would be evaluated together with other RBAC verbs in the auth API handlers.
+
+In particular, this is needed for `SessionRecordingConfig` which has two
+fields:
+* `session_recording` with the possible values: `off`, `node`, `node-sync`,
+  `proxy`, `proxy-sync`;
+* a boolean `proxy_checks_host_keys`.
+
+Some users should be subject to the following restrictions:
+1. The value of `session_recording` can be changed only if its original value
+   is one of `off`, `node` or `node-sync`.
+1. The value of `session_recording` can be changed to only one of `off`, `node`
+   or `node-sync`.
+1. The value of `proxy_checks_host_keys` cannot be changed.
+
+This would be achieved by a new RBAC verb `update.no-proxy`. A user could then
+be assigned a role featuring `update.on-proxy` instead of the regular, more
```

Comment:
nit: let's make these descriptive: `update.proxy_recording_mode` and `update.no_proxy_recording_mode`

Reply:
There was a typo, I meant to introduce only `update.no-proxy` -- now renamed to `update.no_proxy_recording`.